flink-contrib/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBKeyedStateBackend.java

```diff
@@ -1116,13 +1116,27 @@ private void restoreKeyGroupsInStateHandle()
 		 * @throws ClassNotFoundException
 		 * @throws RocksDBException
 		 */
+		@SuppressWarnings("unchecked")
 		private void restoreKVStateMetaData() throws IOException, ClassNotFoundException, RocksDBException {
 
 			KeyedBackendSerializationProxy serializationProxy =
 					new KeyedBackendSerializationProxy(rocksDBKeyedStateBackend.userCodeClassLoader);
 
 			serializationProxy.read(currentStateHandleInView);
 
+			// check for key serializer compatibility; this also reconfigures the
+			// key serializer to be compatible, if it is required and is possible
+			if (StateMigrationUtil.resolveCompatibilityResult(
+					serializationProxy.getKeySerializer(),
+					TypeSerializerSerializationProxy.ClassNotFoundDummyTypeSerializer.class,
+					serializationProxy.getKeySerializerConfigSnapshot(),
+					(TypeSerializer) rocksDBKeyedStateBackend.keySerializer)
+				.isRequiresMigration()) {
+
+				throw new RuntimeException("The new key serializer is not compatible to read previous keys. " +
```

Comment:
In the future, we need to be extra careful in this case. If the key's hash codes change through a state transformation, the their assignment to the operator instances is broken. Maybe we should already leave a comment about this here, so that we remember it in the future.